### PR TITLE
nixos/nftables: extend check redirects

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Description of changes

The `checkPhase` script that runs `nft --check` uses libredirect to link
/etc/services and /etc/protocols inside the sandbox. The problem is that it
resets NIX_REDIRECTS *after* running the user-provided `preCheckRuleset`, so
it's not possible to add more redirects.

This adds the generated /etc/hosts as well, needed to resolve static hostnames,
and a `:$NIX_REDIRECTS` to make this extensible.

###### Things done

- Tested via:
  - [x] `nixosTests.nat.nftables`
  - [x] `nixosTests.firewall`
- [x] Checked the manual with `nix build -f nixos/release.nix manual.x86_64-linux`
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rnhmjoj** [comment]
Btw, if you want to do this now, here's a hack to work around the limitation:
```
export(){
  command export "$1:/etc/hosts=${config.environment.etc.hosts.source}"
}
```

File: nixos/modules/services/networking/nftables.nix [hunk 1]
```diff
@@ -144,7 +150,12 @@ in
           checkPhase = lib.optionalString cfg.checkRuleset ''
             cp $out ruleset.conf
             ${cfg.preCheckRuleset}
-            export NIX_REDIRECTS=/etc/protocols=${pkgs.buildPackages.iana-etc}/etc/protocols:/etc/services=${pkgs.buildPackages.iana-etc}/etc/services
+            export NIX_REDIRECTS=${concatStringsSep ":"
+              [ "/etc/protocols=${pkgs.buildPackages.iana-etc}/etc/protocols"
+                "/etc/services=${pkgs.buildPackages.iana-etc}/etc/services"
+                "/etc/hosts=${config.environment.etc.hosts.source}"
+                "$NIX_REDIRECTS"
+              ]}
```
Comment (Dessix) [inline nixos/modules/services/networking/nftables.nix:158]: Use conditional parameter substitution to extend the list with `NIX_REDIRECTS` if and only if it is assigned with a non-zero-length value.
```suggestion
            export NIX_REDIRECTS="${concatStringsSep ":"
              [ "/etc/protocols=${pkgs.buildPackages.iana-etc}/etc/protocols"
                "/etc/services=${pkgs.buildPackages.iana-etc}/etc/services"
                "/etc/hosts=${config.environment.etc.hosts.source}"
              ]}''${NIX_REDIRECTS:+:''${NIX_REDIRECTS}}"
```

File: nixos/modules/services/networking/nftables.nix [hunk 2]
```diff
@@ -144,7 +150,12 @@ in
           checkPhase = lib.optionalString cfg.checkRuleset ''
             cp $out ruleset.conf
             ${cfg.preCheckRuleset}
-            export NIX_REDIRECTS=/etc/protocols=${pkgs.buildPackages.iana-etc}/etc/protocols:/etc/services=${pkgs.buildPackages.iana-etc}/etc/services
+            export NIX_REDIRECTS=${concatStringsSep ":"
+              [ "/etc/protocols=${pkgs.buildPackages.iana-etc}/etc/protocols"
+                "/etc/services=${pkgs.buildPackages.iana-etc}/etc/services"
+                "/etc/hosts=${config.environment.etc.hosts.source}"
```
Comment (Dessix) [inline nixos/modules/services/networking/nftables.nix:156]: Does non-library redirection actually work with this? I was under the impression `NIX_REDIRECTS` were wholly intended for linking redirects.

**Review by Dessix — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/nftables.nix (hunk 2 above)
Comment (rnhmjoj) [reply nixos/modules/services/networking/nftables.nix:156]: Yeah, it's just hooking the system call wrappers (`open()`, etc.) in libc to rewrite a path.

**Comment by jian-lin** [comment]
related https://github.com/NixOS/nixpkgs/pull/241680

File: nixos/modules/services/networking/nftables.nix (hunk 1 above)
Comment (rnhmjoj) [reply nixos/modules/services/networking/nftables.nix]: Nice trick, implemented.

**Comment by mkg20001** [comment]
if i understand it compared to https://github.com/NixOS/nixpkgs/pull/241680 this allows users to reference paths from the current machine. this doesn't really make sense to me as this build might be building the configuration for another machine that has nothing to do with this one.

when would we want this?

**Comment by rnhmjoj** [comment]
No, it's really the same thing: yes, can *try* to reference any path on the machine, but they won't work because of the sandbox (even if build==host). The paths essentially must be in the store.
The only difference is that in the other PR you have a proper option to define more links, here you do that with a piece of shell script in `preCheckRuleset`.

**Comment by mkg20001** [comment]
ok. in that case i prefer the option, as that can be part of the machine config while in this case it's basically an impure input.

**Comment by rnhmjoj** [comment]
The rational for not having a dedicated option is that probably no one will ever need to add more links, but you never now.
I'm closing in favour of the other PR.

**Pull request closed without merging**